### PR TITLE
Unify loss interface and do not force backpropagation

### PR DESCRIPTION
**Pull request by guillaumekln**
The loss computation had the unwanted side effect of also running the backpropagation. It should not. This PR removes this side effect and makes the loss class expose a single general interface for training and evaluation.

The backward pass in now initiated in `trainer.py` with a simple `loss.backward()`. This will allow other loss manipulation, such as loss scaling for FP16 training. I tested that the loss is the same and memory usage similar in different training configurations: Transformer, RNN, RNN with copy_attn, RNN with TBPTT.

It also adds support of setting `-max_generator_batches 0` to not shard the loss computation (fastest).

### CODE REVIEW

**Comment by pltrdy** [comment]
Doesn't this make the whole sharding idea useless?

I think the idea is precisely to iteratively backward shards for memory efficiency, is that wrong?

**Comment by guillaumekln** [comment]
Thanks for the comment.

From an experimental point of view, the backward pass does require less memory by reducing `-max_generator_batches` (I observed a 2x factor on RNN). I believe it's mostly because the autograd graph is itself sharded and buffers are reused across the generator branches.

That being said it does use more memory than before when sharding is used, about 200MB for a standard Transformer training. I would need more time to really understand the memory requirements in action if that is a concern.

In any case, we should find ways to not run the back-propagation here, it is very limiting.

On a side note, do someone know another toolkit that implements some sort of sharding like this? I'm not sure about the history of this feature in the codebase.

**Comment by jekbradbury** [comment]
The loss sharding was included in the very first WIP version of OpenNMT-Py (in the PyTorch examples directory before the PyTorch public release): https://github.com/pytorch/examples/commit/c048078187945e1defb8abc3246cd305fb038c6a
I believe it dates back to Lua Torch OpenNMT (I certainly used it in my own Torch-based NMT codebase around the same time). Something like it (scheduling ops across the forward and backward pass for lower peak memory use) is automatic in Theano and TensorFlow; it’s straightforward to do this particular optimization manually in Torch, but kind of tricky in Chainer/PyTorch (especially early on).
